packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectQuery.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.realm.internal
+
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.RealmCoreException
+import io.realm.internal.interop.RealmCoreIndexOutOfBoundsException
+import io.realm.internal.interop.RealmCoreInvalidQueryException
+import io.realm.internal.interop.RealmCoreInvalidQueryStringException
+import io.realm.internal.interop.RealmInterop
+import io.realm.query.RealmQuery
+import io.realm.query.RealmScalarQuery
+import io.realm.query.RealmSingleQuery
+import io.realm.query.Sort
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+@Suppress("SpreadOperator")
+internal class RealmObjectQuery<E : RealmObject> constructor(
+    private val realmReference: RealmReference,
+    private val clazz: KClass<E>,
+    private val mediator: Mediator,
+    composedQueryPointer: NativePointer? = null,
+    private val filter: String,
+    private vararg val args: Any?
+) : RealmQuery<E>, Thawable<BaseResults<E>> {
+
+    private val queryPointer: NativePointer = when {
+        composedQueryPointer != null -> composedQueryPointer
+        else -> parseQuery()
+    }
+
+    private val resultsPointer: NativePointer by lazy {
+        RealmInterop.realm_query_find_all(queryPointer)
+    }
+
+    constructor(
+        composedQueryPointer: NativePointer?,
+        objectQuery: RealmObjectQuery<E>
+    ) : this(
+        objectQuery.realmReference,
+        objectQuery.clazz,
+        objectQuery.mediator,
+        composedQueryPointer,
+        objectQuery.filter,
+        *objectQuery.args
+    )
+
+    override fun find(): RealmResults<E> =
+        ElementResults(realmReference, resultsPointer, clazz, mediator)
+
+    override fun query(filter: String, vararg arguments: Any?): RealmQuery<E> {
+        val appendedQuery = tryCatchCoreException {
+            RealmInterop.realm_query_append_query(queryPointer, filter, *arguments)
+        }
+        return RealmObjectQuery(appendedQuery, this)
+    }
+
+    override fun sort(property: String, sortOrder: Sort): RealmQuery<E> =
+        query("TRUEPREDICATE SORT($property ${sortOrder.name})")
```

Comment:
I wonder how that makes the final query look like?

E.g.
```
val desc: String = realm.query<Example>().sort("name", Sort.Descending).getDescription()
// Does this return
// TRUEPREDICATE AND TRUEPREDICATE SORT(name DESC)?
```

That would be a bit annoying, especially since the Core API has methods that allow us to not do that....But It can be fixed in a followup PR.

Reply:
It does 😕  I guess that's another function to be added to the C-API 😓

Reply:
Can we add a TODO or create an issue, then it is fine to resolve  this.